src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs

```diff
@@ -0,0 +1,555 @@
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Sockets.Tests
+{
+    public class SendReceive
+    {
+        private readonly ITestOutputHelper _log;
+
+        public SendReceive(ITestOutputHelper output)
+        {
+            _log = output;
+        }
+
+        private static void SendToRecvFrom_Datagram_UDP(IPAddress leftAddress, IPAddress rightAddress)
+        {
+            const int DatagramSize = 256;
+            const int DatagramsToSend = 256;
+
+            var left = new Socket(leftAddress.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
+            left.BindToAnonymousPort(leftAddress);
+
+            var right = new Socket(rightAddress.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
+            right.BindToAnonymousPort(rightAddress);
+
+            var leftEndpoint = (IPEndPoint)left.LocalEndPoint;
+            var rightEndpoint = (IPEndPoint)right.LocalEndPoint;
+
+            var receiverAck = new ManualResetEventSlim();
+            var senderAck = new ManualResetEventSlim();
+
+            var receivedChecksums = new uint[DatagramsToSend];
+            var leftThread = new Thread(() =>
+            {
+                using (left)
+                {
+                    EndPoint remote = leftEndpoint.Create(leftEndpoint.Serialize());
+                    var recvBuffer = new byte[DatagramSize];
+                    for (int i = 0; i < DatagramsToSend; i++)
+                    {
+                        int received = left.ReceiveFrom(recvBuffer, SocketFlags.None, ref remote);
+                        Assert.Equal(DatagramSize, received);
+                        Assert.Equal(rightEndpoint, remote);
+                        receivedChecksums[(int)recvBuffer[0]] = Fletcher32.Checksum(recvBuffer, 0, received);
+
+                        receiverAck.Set();
+                        senderAck.Wait();
+                        senderAck.Reset();
+                    }
+                }
+            });
+            leftThread.Start();
+
+            var sentChecksums = new uint[DatagramsToSend];
+            using (right)
+            {
+                var random = new Random();
+                var sendBuffer = new byte[DatagramSize];
+                for (int i = 0; i < DatagramsToSend; i++)
+                {
+                    random.NextBytes(sendBuffer);
+                    sendBuffer[0] = (byte)i;
+
+                    int sent = right.SendTo(sendBuffer, SocketFlags.None, leftEndpoint);
+
+                    receiverAck.Wait();
+                    receiverAck.Reset();
+                    senderAck.Set();
+
+                    Assert.Equal(DatagramSize, sent);
+                    sentChecksums[i] = Fletcher32.Checksum(sendBuffer, 0, sent);
+                }
+            }
+
+            leftThread.Join();
```

Comment:
Same as above (throughout the PR): better to use `Join` with a time limit to avoid hanging the CI system.